core/src/main/java/org/elasticsearch/ingest/processor/ScriptProcessor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor;
+
+import org.elasticsearch.ingest.core.*;
+import org.elasticsearch.script.*;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Processor that adds new fields with their corresponding values. If the field is already present, its value
+ * will be replaced with the provided one.
+ */
+public final class ScriptProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "script";
+
+    private final TemplateService.Template template;
+    private final ScriptService scriptService;
+    private final String scriptLang;
+    private final String returnField;
+
+    ScriptProcessor(String tag, TemplateService.Template template, ScriptService scriptService, String scriptLang, String returnField)  {
+        super(tag);
+        this.template = template;
+        this.scriptService = scriptService;
+        this.scriptLang = scriptLang;
+        this.returnField = returnField;
+    }
+
+    @Override
+    public void execute(IngestDocument document) {
+        String rawScript = document.renderTemplate(template);
+        Map<String, Object> vars = new HashMap<>();
+        vars.put("doc", document.getSourceAndMetadata());
+
+        Script script = new Script(rawScript, ScriptService.ScriptType.INLINE, scriptLang, Collections.emptyMap());
+        CompiledScript compiledScript = scriptService.compile(script, ScriptContext.Standard.INGEST, Collections.emptyMap(), null);
```

Comment:
constructing the Script can be done during construction of the processor. The script needs to be recompiled each time if we are to support stored-scripts since the script service needs to fetch the latest from the cluster state

adding stored script support in follow up commit